cpp/src/reductions/scan.cu

```diff
@@ -2,74 +2,149 @@
+                size, stream);
+            RMM_TRY(RMM_ALLOC(&temp_storage, temp_storage_bytes, stream));
+
+            bool const input_has_nulls{ input->valid != nullptr && input->null_count > 0 };
+            if (input_has_nulls) {
+                // allocate temporary column data
+                T* temp_input;
+                RMM_TRY(RMM_ALLOC(&temp_input, size * sizeof(T), stream));
+
+                // copy bitmask
+                size_t valid_byte_length = gdf_get_num_chars_bitmask(size);
+                CUDA_TRY(cudaMemcpyAsync(output->valid, input->valid,
+                        valid_byte_length, cudaMemcpyDeviceToDevice, stream));
+                output->null_count = input->null_count;
 
+                // copy d_input data and replace with 0 if mask is not valid
+                copy_and_replace_nulls(
+                    static_cast<const T*>(input->data), input->valid,
+                    size, temp_input, static_cast<T>(0), stream);
 
-SCAN_IMPL(i8,  int8_t)
-SCAN_IMPL(i32, int32_t)
-SCAN_IMPL(i64, int64_t)
+                // Do scan
+                scan_function(temp_storage, temp_storage_bytes, temp_input,
+                    d_output, size, stream);
 
+                RMM_TRY(RMM_FREE(temp_input, stream));
+            }
+            else {  // Do scan
+                scan_function(temp_storage, temp_storage_bytes,
+                    d_input, d_output, size, stream);
+            }
 
-gdf_error gdf_prefixsum_generic(gdf_column *inp, gdf_column *out,
-                                int inclusive)
+            // Cleanup
+            RMM_TRY(RMM_FREE(temp_storage, stream));
+
+            return GDF_SUCCESS;
+        }
+
+        static
+            gdf_error exclusive_sum(void *&temp_storage, size_t &temp_storage_bytes,
+                const T *input, T *output, size_t size, cudaStream_t stream) {
+            cub::DeviceScan::ExclusiveSum(temp_storage, temp_storage_bytes,
+                input, output, size, stream);
+            CUDA_CHECK_LAST();
+            return GDF_SUCCESS;
+        }
+
+        static
+            gdf_error inclusive_sum(void *&temp_storage, size_t &temp_storage_bytes,
+                const T *input, T *output, size_t size, cudaStream_t stream) {
+            cub::DeviceScan::InclusiveSum(temp_storage, temp_storage_bytes,
+                input, output, size, stream);
+            CUDA_CHECK_LAST();
+            return GDF_SUCCESS;
+        }
+    };
+
+    struct PrefixSumDispatcher {
+        template <typename T,
+            typename std::enable_if_t<std::is_arithmetic<T>::value>* = nullptr>
+            gdf_error operator()(const gdf_column *input, gdf_column *output,
+                bool inclusive, cudaStream_t stream = 0) {
+            GDF_REQUIRE(input->size == output->size, GDF_COLUMN_SIZE_MISMATCH);
+            GDF_REQUIRE(input->dtype == output->dtype, GDF_DTYPE_MISMATCH);
+
+            if (!input->valid) {
```

Comment:
```suggestion
            if (nullptr == input->valid) {
```
Instead of using `!` on pointers, it expresses your intent much clearer to just compare against `nullptr`. 

This is something that will eventually go into our style guide, once we get one...